src/utils/workflow_utils/src/workflow_utils.c

```diff
@@ -1804,57 +1813,133 @@ bool workflow_set_sandbox(ADUC_WorkflowHandle handle, const char* sandbox)
     return true;
 }
 
-// Workfolder =  [root.sandboxfolder]  "/"  ( [parent.workfolder | parent.id]  "/" )+  [handle.workfolder | handle.id]
-char* workflow_get_workfolder(const ADUC_WorkflowHandle handle)
+char* _workflow_copy_config_downloads_folder(const size_t max_size)
 {
-    char dir[1024] = { 0 };
+    char* downloads_folder_path = NULL;
 
-    char* pwf = NULL;
-    char* ret = NULL;
-    char* id = workflow_get_id(handle);
+    const ADUC_ConfigInfo* config = ADUC_ConfigInfo_GetInstance();
+    if (config == NULL)
+    {
+        goto done; /* Config is unitialized or we could not get a reference */
+    }
 
-    // If workfolder explicitly specified, use it.
-    char* wf = workflow_get_string_property(handle, WORKFLOW_PROPERTY_FIELD_WORKFOLDER);
-    if (wf != NULL)
+    size_t download_folder_len = ADUC_StrNLen(config->downloadsFolder, max_size);
+
+    if (download_folder_len == 0 || download_folder_len == max_size)
     {
-        Log_Debug("Property '%s' not NULL - returning cached workfolder '%s'", WORKFLOW_PROPERTY_FIELD_WORKFOLDER, wf);
-        free(id);
-        return wf;
+        Log_Error("Invalid base sandbox dir: '%s'", config->downloadsFolder);
+        goto done;
     }
 
-    // Return ([parent's workfolder] or [default sandbox folder]) + "/" + [workflow id];
+    if (!mallocAndStrcpy_s(&downloads_folder_path, config->downloadsFolder))
+    {
+        goto done;
+    }
+
+done:
+
+    ADUC_ConfigInfo_ReleaseInstance(config);
+
+    return downloads_folder_path;
+}
+
+// Returns the download base directory
+char* workflow_get_root_sandbox_dir(const ADUC_WorkflowHandle handle)
+{
+    char* ret = NULL;
+    char* tempRet = NULL;
+    char* pwf = NULL;
+
     ADUC_WorkflowHandle p = workflow_get_parent(handle);
     if (p != NULL)
     {
         pwf = workflow_get_workfolder(p);
-        sprintf(dir, "%s/%s", pwf, id);
 
-        Log_Debug("Using parent workfolder: '%s/%s'", pwf, id);
+        if (pwf == NULL)
+        {
+            Log_Error("Failed to get parent workfolder");
+            goto done;
+        }
+
+        size_t pwf_len = ADUC_StrNLen(pwf, WORKFLOW_DIR_PATH_MAX_SIZE);
```

Comment:
Does WORKFLOW_DIR_PATH_MAX_SIZE include nul-term? If so, then needs to be -1 here.